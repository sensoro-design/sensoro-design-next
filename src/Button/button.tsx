import React, { useState, useContext, useEffect, ReactNode, forwardRef } from 'react';
import IconLoading from './icon-loading';
import Group from './group';
import classNames from '@pansy/classnames';
import { ConfigContext } from '../ConfigProvider';
import { ButtonProps } from './interface';
import useMergeProps from '../_util/hooks/useMergeProps';

import type { CompoundedComponent } from './interface';

const regexTwoCNChar = /^[\u4e00-\u9fa5]{2}$/;

function processChildren(children?: ReactNode) {
  const childrenList = [];
  let isPrevChildPure = false;
  React.Children.forEach(children, (child) => {
    const isCurrentChildPure = typeof child === 'string' || typeof child === 'number';
    if (isCurrentChildPure && isPrevChildPure) {
      const lastIndex = childrenList.length - 1;
      const lastChild = childrenList[lastIndex];
      childrenList[lastIndex] = `${lastChild}${child}`;
    } else {
      childrenList.push(child);
    }
    isPrevChildPure = isCurrentChildPure;
  });
  return React.Children.map(childrenList, (child) =>
    typeof child === 'string' ? <span>{child}</span> : child
  );
}

const defaultProps: ButtonProps = {
  htmlType: 'button',
  type: 'default',
  shape: 'square',
};

const InternalButton: React.ForwardRefRenderFunction<unknown, ButtonProps> = (baseProps, ref) => {
  const {
    getPrefixCls,
    size: ctxSize,
    autoInsertSpaceInButton,
    componentConfig,
  } = useContext(ConfigContext);
  const props = useMergeProps<ButtonProps>(baseProps, defaultProps, componentConfig?.Button);
  const {
    style,
    className,
    children,
    htmlType,
    bordered,
    type,
    status,
    size,
    shape,
    href,
    anchorProps,
    disabled,
    loading,
    loadingFixedWidth,
    icon,
    iconOnly,
    onClick,
    long,
    ...rest
  } = props;

  const iconNode = loading ? <IconLoading /> : icon;

  const [isTwoCNChar, setIsTwoCNChar] = useState(false);
  const buttonRef = (ref as any) || React.createRef<HTMLElement>();

  useEffect(() => {
    if (autoInsertSpaceInButton && buttonRef && buttonRef.current) {
      const textContent = buttonRef.current.textContent;
      if (regexTwoCNChar.test(textContent)) {
        if (!isTwoCNChar) {
          setIsTwoCNChar(true);
        }
      } else if (isTwoCNChar) {
        setIsTwoCNChar(false);
      }
    }
  }, [buttonRef, autoInsertSpaceInButton]);

  const prefixCls = getPrefixCls('btn');
  const _type = type === 'default' ? 'secondary' : type;
  const classes = classNames(
    prefixCls,
    `${prefixCls}-${_type}`,
    `${prefixCls}-size-${size || ctxSize}`,
    `${prefixCls}-shape-${shape}`,
    {
      [`${prefixCls}-hide-bordered`]: bordered === false && _type === 'dashed',
      [`${prefixCls}-long`]: long,
      [`${prefixCls}-status-${status}`]: status,
      [`${prefixCls}-loading-fixed-width`]: loadingFixedWidth,
      [`${prefixCls}-loading`]: loading,
      [`${prefixCls}-link`]: href,
      [`${prefixCls}-icon-only`]: iconOnly || (!children && children !== 0 && iconNode),
      [`${prefixCls}-disabled`]: disabled,
      [`${prefixCls}-two-chinese-chars`]: isTwoCNChar,
    },
    className
  );

  const handleClick: React.MouseEventHandler<HTMLElement> = (event: any): void => {
    if (loading) {
      typeof event?.preventDefault === 'function' && event.preventDefault();
      return;
    }
    onClick && onClick(event);
  };

  const InnerContent = (
    <>
      {iconNode}
      {processChildren(children)}
    </>
  );

  if (href) {
    const _anchorProps = { ...anchorProps };
    if (disabled) {
      delete _anchorProps.href;
    } else {
      _anchorProps.href = href;
    }
    return (
      <a
        ref={buttonRef}
        {...rest}
        {..._anchorProps}
        style={style}
        className={classes}
        onClick={handleClick}
      >
        {InnerContent}
      </a>
    );
  }

  return (
    <button
      ref={buttonRef}
      {...rest}
      style={style}
      className={classes}
      type={htmlType}
      disabled={disabled}
      onClick={handleClick}
    >
      {InnerContent}
    </button>
  );
}

const Button = forwardRef<unknown, ButtonProps>(InternalButton) as CompoundedComponent;

Button.displayName = 'Button';

Button.__SEN_BUTTON = true;
Button.Group = Group;

export default Button;
