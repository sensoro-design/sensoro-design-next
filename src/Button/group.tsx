import React, { useContext } from 'react';
import classNames from '@pansy/classnames';

import { ConfigContext } from '../ConfigProvider';
import { ButtonGroupProps } from './interface';

function Group(props: ButtonGroupProps, ref) {
  const { className, style, children } = props;
  const { getPrefixCls } = useContext(ConfigContext);
  const prefixCls = getPrefixCls('btn-group');
  const classes = classNames(prefixCls, className);

  return (
    <div ref={ref} className={classes} style={style}>
      {children}
    </div>
  );
}

const GroupComponent = React.forwardRef<unknown, ButtonGroupProps>(Group);

GroupComponent.displayName = 'ButtonGroup';

export default GroupComponent;

export { ButtonGroupProps };
