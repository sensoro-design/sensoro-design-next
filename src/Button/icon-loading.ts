import _extends from "@babel/runtime/helpers/extends";
import _defineProperty from "@babel/runtime/helpers/defineProperty";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

// @ts-ignore
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

import React, { useContext } from 'react';
import { IconContext } from '../../icon/react-icon/context';

function IconLoadingComponent(iconProps, ref) {
  var _useContext = useContext(IconContext),
      prefixCls = _useContext.prefixCls;

  var spin = iconProps.spin,
      className = iconProps.className;

  var props = _objectSpread(_objectSpread({
    ref: ref
  // @ts-ignore
  }, iconProps), {}, {
    className: "".concat(className ? className + ' ' : '').concat(prefixCls, "-icon ").concat(prefixCls, "-icon-loading")
  });

  if (spin) {
    props.className = "".concat(props.className, " ").concat(prefixCls, "-icon-loading");
  }

  delete props.spin;
  delete props.isIcon;
  return /*#__PURE__*/React.createElement(
    "svg", _extends({
      fill: "none",
      viewBox: "0 0 1024 1024",
      width: "1em",
      height: "1em"
    }, props), /*#__PURE__*/
    React.createElement(
      "defs",
      {},
      React.createElement('linearGradient',
        {
          id:"left",
          x1: "100%",
          y1:"80%",
          x2: "100%",
          y2:"0%",
        },
        React.createElement('stop', {
          offset:"0%", stopColor:"currentColor", stopOpacity:"0.5"
        }),
        React.createElement('stop', {
          offset: "100%", stopColor: "currentColor", stopOpacity: "1"
        }
      )),
      React.createElement('linearGradient',
        {
          id: "right", x1: "0%", y1: "0%", x2: "0%", y2: "80%"
        },
        React.createElement('stop', {
          offset:"0%", stopColor:"currentColor", stopOpacity:"0"
        }),
        React.createElement('stop', {
          offset: "100%", stopColor: "currentColor", stopOpacity: "0.5"
        }
      ))
    ),
    React.createElement("path", {
      fill: "url(#left)",
      d: "M512.00000004 64v76.032a369.92 369.92 0 0 0-263.04 108.864A370.432 370.432 0 0 0 140.03200004 512a369.92 369.92 0 0 0 108.928 263.04A370.432 370.432 0 0 0 512.00000004 883.968V960A448 448 0 0 1 512.00000004 64z"
    }),
    React.createElement("path", {
      fill: "url(#right)",
      d: "M959.99999996 512a448 448 0 0 1-432.64 447.744L511.99999996 960v-76.032c50.304 0 99.008-9.792 144.64-29.184a372.032 372.032 0 0 0 118.4-79.808A369.088 369.088 0 0 0 883.96799996 512L959.99999996 512.064z"
    })
  );
}

var IconLoading = /*#__PURE__*/React.forwardRef(IconLoadingComponent);

IconLoading.defaultProps = {
  isIcon: true
};
IconLoading.displayName = 'IconLoading';
export default IconLoading;
