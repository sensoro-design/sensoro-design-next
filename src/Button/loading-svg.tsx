// This icon file is generated automatically
import React from 'react';
import uniqueId from "lodash/uniqueId";

import type { CustomIconComponentProps } from '@sensoro-design/icons/lib/components/Icon';

export const LoadingGradient = (props: CustomIconComponentProps) => {
  const id = uniqueId();

  return (
    <svg
      {...props}
      xmlns="http://www.w3.org/2000/svg"
      xmlnsXlink="http://www.w3.org/1999/xlink"
      viewBox="0 0 1024 1024"
    >
      <defs>
        <linearGradient id={`${id}-left`} x1="100%" y1="80%" x2="100%" y2="0%">
          <stop offset="0%" stopColor="currentColor" stopOpacity="0.5"/>
          <stop offset="100%" stopColor="currentColor" stopOpacity="1"/>
        </linearGradient>
        <linearGradient id={`${id}-right`} x1="0%" y1="0%" x2="0%" y2="80%">
          <stop offset="0%" stopColor="currentColor" stopOpacity="0"/>
          <stop offset="100%" stopColor="currentColor" stopOpacity="0.5"/>
        </linearGradient>
      </defs>
      <path fill={`url(#${id}-left)`} d="M512.00000004 64v76.032a369.92 369.92 0 0 0-263.04 108.864A370.432 370.432 0 0 0 140.03200004 512a369.92 369.92 0 0 0 108.928 263.04A370.432 370.432 0 0 0 512.00000004 883.968V960A448 448 0 0 1 512.00000004 64z" />
      <path fill={`url(#${id}-right)`} d="M959.99999996 512a448 448 0 0 1-432.64 447.744L511.99999996 960v-76.032c50.304 0 99.008-9.792 144.64-29.184a372.032 372.032 0 0 0 118.4-79.808A369.088 369.088 0 0 0 883.96799996 512L959.99999996 512.064z" />
    </svg>
  )
};
