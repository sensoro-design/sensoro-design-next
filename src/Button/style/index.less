@import '../../style/theme/default.less';
@import './token.less';
@import './mixin.less';

.@{btn-prefix-cls} {
  display: inline-block;
  position: relative;
  outline: none;
  font-weight: @btn-font-weight;
  appearance: none;
  user-select: none;
  cursor: pointer;
  white-space: nowrap;
  transition: all @transition-duration-1 @transition-timing-function-linear;
  box-sizing: border-box;
  line-height: @line-height-base;

  > a:only-child {
    color: currentColor;
  }

  &:active {
    transition: none;
  }

  &:empty {
    display: inline-block;
    vertical-align: bottom;
  }

  &-long {
    display: block;
    width: 100%;
  }

  // link
  &-link {
    display: inline-flex;
    align-items: center;
    justify-content: center;
    text-decoration: none;

    &:not([href]) {
      color: var(~'@{arco-cssvars-prefix}-color-text-4');
    }

    &:hover {
      text-decoration: none;
    }
  }

  &-loading {
    cursor: default;
    position: relative;

    &::before {
      content: '';
      position: absolute;
      top: -1px;
      right: -1px;
      bottom: -1px;
      left: -1px;
      z-index: 1;
      display: block;
      background: #fff;
      border-radius: inherit;
      opacity: 0.4;
      transition: opacity @transition-duration-1 @transition-timing-function-linear;
      pointer-events: none;
    }
  }

  &-loading-fixed-width {
    transition: none;
  }

  &-two-chinese-chars > *:not(svg) {
    letter-spacing: 0.3em;
    margin-right: -0.3em;
  }

  a&-icon-only {
    display: inline-flex;
    align-items: center;
    justify-content: center;
    vertical-align: top;
  }
}

.btn-type(outline);
.btn-status(outline, warning);
.btn-status(outline, danger);
.btn-status(outline, success);

.btn-type(primary);
.btn-status(primary, warning);
.btn-status(primary, danger);
.btn-status(primary, success);

.btn-type(secondary);
.btn-status(secondary, warning);
.btn-status(secondary, danger);
.btn-status(secondary, success);

.btn-type(dashed);
.btn-status(dashed, warning);
.btn-status(dashed, danger);
.btn-status(dashed, success);

.btn-type(text);
.btn-status(text, warning);
.btn-status(text, danger);
.btn-status(text, success);

.btn-size(small);
.btn-size(default);
.btn-size(large);

.@{btn-prefix-cls}-group {
  display: inline-block;

  .@{btn-prefix-cls}-outline:not(:first-child),
  .@{btn-prefix-cls}-dashed:not(:first-child) {
    margin-left: -1px;
  }

  .@{btn-prefix-cls}-primary:not(:last-child) {
    border-right: 1px solid @btn-primary-color-bg_hover;
  }

  .@{btn-prefix-cls}-secondary:not(:last-child) {
    border-right: 1px solid @btn-secondary-color-bg_hover;
  }

  .@{btn-prefix-cls}-status-warning:not(:last-child) {
    border-right: 1px solid @btn-primary-color-bg_warning_hover;
  }

  .@{btn-prefix-cls}-status-danger:not(:last-child) {
    border-right: 1px solid @btn-primary-color-bg_danger_hover;
  }

  .@{btn-prefix-cls}-status-success:not(:last-child) {
    border-right: 1px solid @btn-primary-color-bg_success_hover;
  }

  .@{btn-prefix-cls}-outline,
  .@{btn-prefix-cls}-dashed {
    &:hover,
    &:active {
      z-index: 2;
    }
  }

  .@{btn-prefix-cls}:first-child {
    border-top-right-radius: 0;
    border-bottom-right-radius: 0;
  }

  .@{btn-prefix-cls}:last-child {
    border-top-left-radius: 0;
    border-bottom-left-radius: 0;
  }

  .@{btn-prefix-cls}:not(:first-child):not(:last-child) {
    border-radius: 0;
  }
}

@{arco-theme-tag} {
  &[arco-theme='dark'] {
    .@{btn-prefix-cls}-primary.@{btn-prefix-cls}-disabled {
      color: rgba(255, 255, 255, 30%);
    }
  }
}
