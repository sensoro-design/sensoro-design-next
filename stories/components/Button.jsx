import React from 'react';
import { Button } from '@self';
import { IconInfoCircleFill, IconExclamationCircleFill, IconCloseCircleFill } from '@self/icon';

class DemoButton extends React.Component {
  render() {
    return (
      <div>
        <Button style={{ margin: 20 }}>默认按钮</Button>
        <Button onClick={(event) => console.log(event)} type="primary" style={{ margin: 20 }}>
          <IconInfoCircleFill />
          主要按钮
        </Button>
        <Button type="warning" style={{ margin: 20 }}>
          <IconExclamationCircleFill />
          危险按钮
        </Button>
        <Button type="danger" style={{ margin: 20 }}>
          <IconCloseCircleFill />
          高危按钮
        </Button>
        <Button disabled type="primary" style={{ margin: 20 }}>
          失效按钮
        </Button>
        <Button type="dashed" style={{ margin: 20 }}>
          虚线按钮
        </Button>
        <Button type="text" style={{ margin: 20 }}>
          文字按钮
        </Button>
        <Button size="small" loading style={{ margin: 20 }} onClick={() => console.log('!!!!!')}>
          Default
        </Button>
        <Button size="large" loading style={{ margin: 20 }} onClick={() => console.log('~~~~')}>
          Default
        </Button>
        <Button size="large" type="primary" loading style={{ margin: 20 }}>
          Default
        </Button>
        <Button size="large" type="warning" loading style={{ margin: 20 }}>
          Default
        </Button>
        <Button size="large" type="danger" loading style={{ margin: 20 }}>
          Default
        </Button>
        <Button size="large" type="dashed" loading style={{ margin: 20 }}>
          Default
        </Button>
        <Button size="large" type="text" loading style={{ margin: 20 }}>
          Default
        </Button>
      </div>
    );
  }
}

export default DemoButton;
