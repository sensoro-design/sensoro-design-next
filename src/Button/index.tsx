import Button from './button';

export type {
  ButtonProps,
  ButtonGroupProps,
  ButtonType,
  ButtonStatus,
} from './interface';

export default Button;
