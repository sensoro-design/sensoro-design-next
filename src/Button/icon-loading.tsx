import React from 'react';
import classNames from '@pansy/classnames';

import Icon from '@sensoro-design/icons/lib/components/Icon';
import { LoadingGradient } from './loading-svg';

import type { CustomIconComponentProps } from '@sensoro-design/icons/es/components/Icon';

export default (props: Partial<CustomIconComponentProps>) => {
  return <Icon {...props} className={classNames('s-icon-loading', props.className)} component={LoadingGradient} />;
};
