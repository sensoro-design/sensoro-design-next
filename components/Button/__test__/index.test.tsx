import React from 'react';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import mountTest from '../../../tests/mountTest';
import componentConfigTest from '../../../tests/componentConfigTest';
import Button from '..';

mountTest(Button);
componentConfigTest(Button, 'Button');

describe('button', () => {
  it('click callback correctly', () => {
    const mockFn = jest.fn();
    const component = mount(<Button onClick={mockFn} />);
    const button = component.find('button');
    button.simulate('click');
    const mockFnCallLength = mockFn.mock.calls.length;
    expect(mockFnCallLength).toBe(1);

    act(() => {
      component.setProps({
        disabled: true,
      });
    });

    button.simulate('click');
    expect(mockFn.mock.calls.length).toBe(mockFnCallLength);
  });

  it('render multiple children correctly', () => {
    const button = mount(
      <Button>
        1{'  '}2{'  '}3{'  '}
      </Button>
    );
    expect(button.text()).toEqual('1  2  3  ');
  });

  it('use context autoInsertSpaceInButton correctly', () => {
    const mockText = '测试';
    const realContext = React.useContext;
    const mockContext = jest.fn().mockReturnValue({
      autoInsertSpaceInButton: true,
      getPrefixCls: (componentName) => `test-${componentName}`,
    });

    const prefixCls = `.test-btn`;
    React.useContext = mockContext;

    const button = mount(<Button>{mockText}</Button>);
    expect(button.find(prefixCls).hasClass(`test-btn-two-chinese-chars`)).toBe(true);

    React.useContext = realContext;
  });

  it('render href type correctly', () => {
    const button = mount(
      <Button type="primary" href="https://arco.design">
        测试
      </Button>
    );
    expect(button.find('button')).toHaveLength(0);
    expect(button.find('a')).toHaveLength(1);
    expect(button.find('a').prop('href')).not.toBeUndefined();

    act(() => {
      button.setProps({
        disabled: true,
      });
    });
    expect(button.find('a').prop('href')).toBeUndefined();
  });
});
