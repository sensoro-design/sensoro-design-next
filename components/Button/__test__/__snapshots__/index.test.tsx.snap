// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigProvider componentConfig.Button default 1`] = `
<button
  class="arco-btn arco-btn-secondary arco-btn-size-default arco-btn-shape-square"
  type="button"
/>
`;

exports[`ConfigProvider componentConfig.Button set className globally 1`] = `
<button
  class="arco-btn arco-btn-secondary arco-btn-size-default arco-btn-shape-square global-default-class"
  type="button"
/>
`;

exports[`ConfigProvider componentConfig.Button set className globally and component set className itself. 1`] = `
<button
  class="arco-btn arco-btn-secondary arco-btn-size-default arco-btn-shape-square component-class"
  type="button"
/>
`;
